### ISSUE TITLE
Blazor wasm .NET 6 rc2 unable to use RegisterForJavaScript RootComponents when publishing

### ISSUE THREAD

**Issue by corylanza** [dotnet/aspnetcore#38044]
### Describe the bug
When building and running the Blazor wasm project locally I am able to use the new `builder.RootComponents.RegisterForJavaScript` functionality without issue, but when publishing to a folder and uploading the published project, I am getting an exception when I try to add root component via JavaScript stating that `AddRootComponent` is not a public invokable method with `JSInvokableAttribute`. Worth noting that I see no other issues and when using `builder.RootComponents.Add`, it seems to be only when attempting to add root components via JavaScript even after they have been registered successfully and only when the project is published to a folder.

### To Reproduce
[Example repo](https://github.com/corylanza/Net6RegisterForJavaScriptTest)

[Live Demo hosted on Github pages](https://corylanza.github.io/Net6RegisterForJavaScriptTest/) (error can be seen in browser console)

in Program.cs
`builder.RootComponents.RegisterForJavaScript<BlazorApp>("blazor-app");`
Executed on page after Blazor.Start
`await Blazor.rootComponents.add(element, 'blazor-app', {  });`

### Exceptions (if any)
`Uncaught (in promise) Error: System.ArgumentException: The type 'WebRendererInteropMethods' does not contain a public invokable method with [JSInvokableAttribute("AddRootComponent")].
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.GetCachedMethodInfo(IDotNetObjectReference , String )
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.InvokeSynchronously(JSRuntime , DotNetInvocationInfo& , IDotNetObjectReference , String )
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.BeginInvokeDotNet(JSRuntime , DotNetInvocationInfo , String )`

### Further technical details
- ASP.NET Core version: Microsoft.AspNetCore.Components.WebAssembly" Version="6.0.0-rc.2.*
- VS 2022 RC 7
- Include the output of `dotnet --info`:

<details>
  <summary>dotnet --info Output</summary>

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-rc.2.21505.57
 Commit:    ab39070116

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rc.2.21505.57\

Host (useful for support):
  Version: 6.0.0-rc.2.21480.5
  Commit:  6b11d64e7e

.NET SDKs installed:
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.1.802 [C:\Program Files\dotnet\sdk]
  3.0.100-preview5-011568 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]
  5.0.301 [C:\Program Files\dotnet\sdk]
  5.0.402 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview5-19227-01 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview5-27626-15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview5-27626-15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

</details>

**Comment by SteveSandersonMS** [comment]
I'm wondering if this is a caching issue. What happens if you visit the published site in a different browser?

**Comment by corylanza** [comment]
@SteveSandersonMS Same deal, I have tried guest windows / alternate browsers and also force deleting the old files to ensure that the compiled dlls are updated

**Comment by SteveSandersonMS** [comment]
Could you also try deleting the bin/obj folders from your project and the re-publishing? Maybe it's a build incrementalism issue.

**Comment by corylanza** [comment]
I tried deleting bin/obj folders as well as clearing NuGet Cache/restoring and still see the same error. Worth noting that other .NET 6 new Blazor features are working and the published dlls (e.g. `Microsoft.AspNetCore.Components.WebAssembly.dll`) have file version `6.0.21.48010` / product version `6.0.0-rc.2.21480`... so it seems like its not a caching issue

**Comment by SteveSandersonMS** [comment]
OK, thanks for the additional details. This is surprising to me because I've tried this in another app and the registered JavaScript components worked fine after publishing.

Would you be able to create a minimal repro app as a GitHub repo? That is, what's the smallest set of changes you need to make to the default project template to show this issue? We could then investigate. Thanks!

**Comment by corylanza** [comment]
@SteveSandersonMS I've created a minimal repo [here](https://github.com/corylanza/Net6RegisterForJavaScriptTest/tree/master) with a live demo of the error hosted on github pages [here](https://corylanza.github.io/Net6RegisterForJavaScriptTest/). In browser console you should see the exception as originally posted
![image](https://user-images.githubusercontent.com/24576321/140174743-895343d0-bfff-4706-bdf0-ac5b4bbe214d.png)

**Comment by da1rren** [comment]
We have the same issue in the .net 6.0 release version.

**Comment by Xenoage** [comment]
Same issue when publishing the [demo app](https://github.com/aspnet/samples/tree/main/samples/aspnetcore/blazor/JSComponentGeneration) for Angular component creation presented in the [.NET Conf 2021](https://www.youtube.com/watch?v=GKu-vRxOWr8) talk.

**Comment by Xenoage** [comment]
Interestingly, the problem does _not_ occur when calling from a custom initializer JS function, by using the `javaScriptInitializer` parameter in `RegisterForJavaScript`. See the [changes](https://github.com/Xenoage/Net6RegisterForJavaScriptTest/commit/b9212a95bbc4d8f2d1db669e84b3a5ba64fc7618) in my [fork](https://github.com/Xenoage/Net6RegisterForJavaScriptTest) of @corylanza 's minimal demo; compiled demo running [here](https://xenoage.github.io/Net6RegisterForJavaScriptTest/) without problems.

**Comment by corylanza** [comment]
This is a great workaround, thank you @Xenoage

**Comment by Xenoage** [comment]
Unfortunately, this workaround does not help in the Angular demo for example. [This call](https://github.com/aspnet/samples/blob/716ddf3c9c769acce9ad9799ce5751b506c6f877/samples/aspnetcore/blazor/JSComponentGeneration/angular-app-with-blazor/src/app/blazor-adapter/blazor-adapter.component.ts#L54) is made when an Angular component is added, and I do not see a way how I could "combine" it with the `javaScriptInitializer`.

**Comment by Xenoage** [comment]
OK, I got the Angular example working, using the following weird hack: I added a `javaScriptInitializer: "initHack"` to the `RegisterForJavaScript` call on Blazor side. Then I added the following function to `index.html` to receive the call and remember the current value of `Blazor.rootComponents` globally in the `window` object: `window.initHack= function () { window.rootComponentsHack = Blazor.rootComponents; };`. Then, in `blazor-adapter.component.ts`, I modified the line where the component is added: `this._addComponentPromise = /*BEFORE: Blazor.rootComponents  NOW: */(window as any).rootComponentsHack.add(nativeElement, ...` - and this works!
Of course this is nothing I want to use in production. Maybe it helps finding the problem.

**Comment by Xenoage** [comment]
I also got the Angular example working, and I think I have found the problem. The error occurs when published when `RegisterForJavaScript` is called without the `javaScriptInitializer` parameter! As soon as the parameter is set in Blazor, e.g. `configuration.RegisterForJavaScript<MyComponent>("my-component", javaScriptInitializer: "myInitMethod");`, it works. It does not even matter if `myInitMethod` is implemented on the website at all.

**Comment by Xenoage** [comment]
@SteveSandersonMS I'm really not familiar with the ASP.NET Core / Blazor codebase, just an idea what could be the cause of the problem. Look at the definiton of 2-parameter [JSComponentConfigurationStore.Add](https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Components/Web/src/JSComponents/JSComponentConfigurationStore.cs#L27) . Like its [3-parameter brother](https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Components/Web/src/JSComponents/JSComponentConfigurationStore.cs#L45), this method is called from `JSComponentConfigurationExtensions`. The 3-parameter version has `DynamicDependency` attributes, the 2-parameter version not.

This would explain, why the error `AddRootComponent is not a public invokable method` occurs when the 2-parameter version (i.e. without the `javaScriptInitializer` parameter) is called.

**Comment by mkArtakMSFT** [comment]
Thanks for your patience. Could you please confirm that this is fixed in the latest 6.0 release?
